pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go

```diff
@@ -75,116 +77,173 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
-			Message: strings.Join(progressingMessages, "\n"),
-		})
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:   "APIServerDaemonSetAvailable",
+				Status: operatorv1.ConditionTrue,
+				Reason: "AsExpected",
+			})))...,
+		)
 	}
 
-	// TODO this is changing too early and it was before too.
-	operatorConfig.Status.ObservedGeneration = operatorConfig.ObjectMeta.Generation
-	resourcemerge.SetDaemonSetGeneration(&operatorConfig.Status.Generations, actualDaemonSet)
-	if len(errors) > 0 {
-		message := ""
-		for _, err := range errors {
-			message = message + err.Error() + "\n"
-		}
-		v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions, operatorv1.OperatorCondition{
-			Type:    workloadDegradedCondition,
-			Status:  operatorv1.ConditionTrue,
-			Message: message,
-			Reason:  "SyncError",
-		})
+	// If the daemonset is up to date and the operatorConfig are up to date, then we are no longer progressing
+	daemonSetAtHighestGeneration := actualDaemonSet.ObjectMeta.Generation == actualDaemonSet.Status.ObservedGeneration
+	if !daemonSetAtHighestGeneration {
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetProgressing",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "NewGeneration",
+				Message: fmt.Sprintf("daemonset/apiserver.openshift-operator: observed generation is %d, desired generation is %d.", actualDaemonSet.Status.ObservedGeneration, actualDaemonSet.ObjectMeta.Generation),
+			})))...,
+		)
 	} else {
-		v1helpers.SetOperatorCondition(&operatorConfig.Status.Conditions, operatorv1.OperatorCondition{
-			Type:   workloadDegradedCondition,
-			Status: operatorv1.ConditionFalse,
-		})
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:   "APIServerDaemonSetProgressing",
+				Status: operatorv1.ConditionFalse,
+				Reason: "AsExpected",
+			})))...,
+		)
 	}
 
-	// if we are available, we need to try to set our versions correctly.
-	if v1helpers.IsOperatorConditionTrue(operatorConfig.Status.Conditions, operatorv1.OperatorStatusTypeAvailable) {
+	// TODO this is changing too early and it was before too.
+	errors = append(errors,
+		appendErrors(v1helpers.UpdateStatus(c.operatorClient, func(status *operatorv1.OperatorStatus) error {
+			status.ObservedGeneration = operatorConfig.ObjectMeta.Generation
+			return nil
+		}))...,
+	)
+	errors = append(errors,
+		appendErrors(v1helpers.UpdateStatus(c.operatorClient, func(status *operatorv1.OperatorStatus) error {
+			resourcemerge.SetDaemonSetGeneration(&status.Generations, actualDaemonSet)
+			return nil
+		}))...,
+	)
+
+	daemonSetHasAllPodsAvailable := actualDaemonSet.Status.NumberAvailable == actualDaemonSet.Status.DesiredNumberScheduled
```

Comment:
what's the difference between `NumberAvailable ` and `UpdatedNumberScheduled`?

Reply:
This is consistent with `kubectl` rollout status:

```go
112:		if daemon.Status.NumberAvailable < daemon.Status.DesiredNumberScheduled {
113:			return fmt.Sprintf("Waiting for daemon set %q rollout to finish: %d of %d updated pods are available...\n", daemon.Name, daemon.Status.NumberAvailable, daemon.Status.DesiredNumberScheduled), false, nil
```